R/SandwichLayerVariance.R

```diff
@@ -558,3 +559,459 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
+      vmat <- .get_DB_wo_covadj_se(x)
+    }
+  }
+  name <- colnames(x$model)[grep("z", colnames(x$model))]
+  colnames(vmat) <- name
+  rownames(vmat) <- name
+  
+  attr(vmat, "type") <- "DB0"
+  return(vmat)
+}
+
+#' @title (Internal) Design-based variance for models with covariance adjustment
+#' @param x a fitted \code{teeMod} model
+#' @details Calculate bread matrix for design-based variance estimate for 
+#'  \code{teeMod} models with covariance adjustment and without absorbed effects
+#' @return design-based variance estimate of the main treatment effect estimate
+#' @keywords internal
+.get_DB_covadj_se <- function(x, ...){
+  if (x@absorbed_intercepts) {
+    stop("x should not have absorbed intercepts")
+  }
+  design_obj <- x@Design
+  name_trt <- colnames(design_obj@structure)[design_obj@column_index == "t"]
+  if (!is.na(x$call$offset@fitted_covariance_model$coefficients[name_trt])) {
+    stop(paste("Design-based standard errors cannot be calculated for",
+               "tee models with treatment in prior covariance adjustment"))
+  }
+  
+  bread <- .get_DB_covadj_bread(x)
+  
+  signs1 <- ifelse(t(t(bread$b1[2,])) %*% bread$b1[2,] > 0, 1, 0)
+  signs2 <- ifelse(t(t(bread$b1[2,])) %*% bread$b2[2,] > 0, 1, 0)
+  signs3 <- ifelse(t(t(bread$b2[2,])) %*% bread$b2[2,] > 0, 1, 0)
+  
+  meat <- .get_DB_covadj_meat(x)
+  
+  term1 <- bread$b1 %*% (meat$m1u * signs1 + meat$m1l * (1-signs1)) %*% t(bread$b1)
+  term2 <- bread$b2 %*% t(meat$m2u * signs2 + meat$m2l * (1-signs2)) %*% t(bread$b1)
+  term3 <- bread$b2 %*% (meat$m3u * signs3 + meat$m3l * (1-signs3)) %*% t(bread$b2)
+  
+  vmat <- term1 + 2*term2 + term3
+  return(as.matrix(vmat[2,2]))
+}
+
+#' @title (Internal) Bread matrix of design-based variance
+#' @param x a fitted \code{teeMod} model
+#' @details Calculate bread matrix for design-based variance estimate for 
+#'  \code{teeMod} models with covariance adjustment and without absorbed effects
+#' @return a list of bread matrices
+#' @keywords internal
+.get_DB_covadj_bread <- function(x, ...) {
+  a11inv <- .get_a11_inverse(x)
+  a21 <- .get_a21(x)
+  a22inv <- .get_a22_inverse(x)
+  a21 <- matrix(0, nrow = nrow(a21), ncol = ncol(a21))
+  a22inv <- matrix(c(1,0,-1,1), nrow = 2, byrow = TRUE)
+  C <- matrix(c(1,0,0,1), nrow = 2, byrow = TRUE)
+  
+  n <- nrow(x$model)
+  bread1 <- a22inv %*% a21 %*% a11inv / n
+  bread2 <- - a22inv %*% C / n
+  
+  return(list(b1 = bread1, b2 = bread2))
+}
+
+#' @title (Internal) Meat matrix of design-based variance
+#' @param x a fitted \code{teeMod} model
+#' @details Calculate upper and lower bound estimates of meat matrix 
+#' for design-based variance estimate for \code{teeMod} models 
+#' with covariance adjustment and without absorbed effects
+#' @return a list of meat matrix bounds
+#' @keywords internal
+.get_DB_covadj_meat <- function(x, ...) {
+  agg <- .aggregate_to_cluster(x)
+  data <- agg[[1]]
+  bid <- data[, agg[[2]]]  # block ids
+  zobs <- data[, agg[[3]]] # observed zs
```

Comment:
Instead of 1, 2, and 3, using the names of `agg` will make it a little clearer what `.aggregate_to_cluster()` returns

Reply:
Updated in commit https://github.com/benbhansen-stats/propertee/pull/185/commits/6eb8927664674088f61b3323683e7d592a2de431